nengo_loihi/snips/nengo_io.c.template

```diff
@@ -17,41 +25,100 @@ void nengo_io(runState *s) {
     int outChannel = getChannelID("nengo_io_c2h");
     int32_t count[1];
-    int32_t spike[2];
+    int32_t spike[3];
     int32_t error[2];
     int32_t output[N_OUTPUTS];
 
+    int32_t *value = (int32_t*)(s->userData+IDX_VALUES);
+    int32_t *pos_accum = (int32_t*)(s->userData+IDX_POS_ACCUMULATORS);
+    int32_t *neg_accum = (int32_t*)(s->userData+IDX_NEG_ACCUMULATORS);
+
     if (inChannel == -1 || outChannel == -1) {
         printf("Got an invalid channel ID\n");
         return;
     }
 
-    if (s->time % 100 == 0) {
+    if (s->time == 1) {
+        printf("initializing\n");
+        for (int i=0; i<N_INPUTS; i++) {
+            readChannel(inChannel, spike, 3);
+            s->userData[IDX_SPIKE_TARGETS+(i*3)+0] = spike[0];  // core id
+            s->userData[IDX_SPIKE_TARGETS+(i*3)+1] = spike[1];  // pos axon
+            s->userData[IDX_SPIKE_TARGETS+(i*3)+2] = spike[2];  // neg axon
+            printf("  spike target %d: (%d %d %d)\n",
+                   i, spike[0], spike[1], spike[2]);
+        }
+    }
+
+    if (s->time % 1000 == 0) {
         printf("time %d\n", s->time);
     }
 
-    readChannel(inChannel, count, 1);
+    //readChannel(inChannel, count, 1);
     // printf("count %d\n", count[0]);
 
-    for (int i=0; i < count[0]; i++) {
-        readChannel(inChannel, spike, 2);
-        // printf("send spike %d.%d\n", spike[0], spike[1]);
-        coreId = (CoreId) { .id=spike[0] };
-        nx_send_discrete_spike(s->time, coreId, spike[1]);
+    // --- read input values (once every IO_STEPS)
+    // Note that we do this at the *start* of an IO_STEP period (since
+    //  s->time starts at 1), so that we don't have a period at the beginning
+    //  of the simulation where we are ignoring the input.
+    if ((IO_STEPS==1) || (s->time % IO_STEPS == 1)) {
+        for (int i=0; i < N_INPUTS; i++) {
+            readChannel(inChannel, spike, 1);
+            //printf("  %d: stim value %d.%d\n", s->time, i, spike[0]);
+            value[i] = spike[0];  // discretized version of the real value
+                                  //  to be used as input
+        }
+
+        // Communicate with learning snip
+        for (int i=0; i < N_ERRORS; i++) {
+            readChannel(inChannel, error, 2);
+            // printf("send error %d.%d\n", error[0], error[1]);
+            s->userData[0] = error[0];
+            s->userData[1] = error[1];
+        }
     }
 
-    // Communicate with learning snip
-    for (int i=0; i < N_ERRORS; i++) {
-        readChannel(inChannel, error, 2);
-        // printf("send error %d.%d\n", error[0], error[1]);
-        s->userData[0] = error[0];
-        s->userData[1] = error[1];
+    // --- turn input values into spikes using accumulators
+    for (int i=0; i < N_INPUTS; i++) {
+        //printf("%d   value:%d  accum:%d\n");
+        pos_accum[i] += (value[i] + (1<<15));
+        if (pos_accum[i] >= (1<<16)) {
+            uint8_t core = s->userData[IDX_SPIKE_TARGETS+(i*3)];
```

Comment:
Can you move all variable declarations to the beginning? I think it's a pretty old school C thing to do, but it's what I'm used to.

Reply:
:)  Sounds fair.  :)

Reply:
Yeah, I think that's back from the days before compilers were very good at optimizing things. Personally, I think readability is the most important factor. It can also help with debugging to have your variables only declared in the scope you want to use them in, so you don't accidentally use them out of scope (one of the reasons old-school C programs were not fun to debug).

Reply:
Yeah I've not done enough C stuff in a professional setting to know what the standards are nowadays, but perhaps we could adopt [PEP-7](https://www.python.org/dev/peps/pep-0007/) since we're using PEP-8.

Reply:
PEP-7 sounds good to me....  that's pretty close to what we've already been doing, too, other than cuddled else statements and the brace that starts a function....